metricflow/model/semantics/join_validator.py

```diff
@@ -58,34 +75,140 @@ class DataSourceJoinValidator:
+        for identifier in data_source.identifiers:
+            # Verify single-hop joins
+            identifier_reference = IdentifierReference(element_name=identifier.name)
+            joinable_data_source_references = self._get_new_joinable_data_sources_for_identifier(
+                left_data_source_reference=left_data_source_reference,
+                known_data_source_joins=data_source_joins,
+                identifier_reference=identifier_reference,
+            )
+            for (joinable_data_source_reference, join_type) in joinable_data_source_references:
+                data_source_joins[joinable_data_source_reference.data_source_name] = DataSourceJoin(
+                    left_data_source_reference=left_data_source_reference,
+                    right_data_source_reference=joinable_data_source_reference,
+                    identifier_joins=[
+                        DataSourceIdentifierJoin(identifier_reference=identifier_reference, join_type=join_type)
+                    ],
+                )
+
+                if not include_multi_hop:
+                    continue
+
+                # Verify multi-hop joins
+                joinable_data_source = self._data_source_semantics.get_by_reference(
+                    data_source_reference=joinable_data_source_reference
+                )
+                assert joinable_data_source is not None
+                for secondary_identifier in joinable_data_source.identifiers:
+                    secondary_identifier_reference = IdentifierReference(element_name=secondary_identifier.name)
+                    secondary_joinable_data_source_references = self._get_new_joinable_data_sources_for_identifier(
+                        left_data_source_reference=joinable_data_source_reference,
+                        known_data_source_joins=data_source_joins,
+                        identifier_reference=secondary_identifier_reference,
+                    )
+                    for (
+                        secondary_joinable_data_source_reference,
+                        secondary_join_type,
+                    ) in secondary_joinable_data_source_references:
+                        if (
+                            secondary_joinable_data_source_reference.data_source_name
+                            == left_data_source_reference.data_source_name
+                        ):
+                            continue
+                        data_source_joins[secondary_joinable_data_source_reference.data_source_name] = DataSourceJoin(
+                            left_data_source_reference=left_data_source_reference,
+                            right_data_source_reference=secondary_joinable_data_source_reference,
+                            identifier_joins=[
+                                DataSourceIdentifierJoin(
+                                    identifier_reference=identifier_reference, join_type=join_type
+                                ),
```

Comment:
Do we need to keep the intermediate data source reference for this?

Reply:
Hmmm good question.